File: app/javascript/channels/query_channel.js
import consumer from "./consumer"

export const createQueryChannel = (queryId) => {
  return consumer.subscriptions.create(
    { channel: "QueryChannel", query_id: queryId },
    {
      connected() {
        console.log(" Connected to QueryChannel")
        this.isConnected = true
      },

      disconnected() {
        console.log("Disconnected from QueryChannel")
        this.isConnected = false
      },

      received(data) {  
        const chatBox = document.getElementById("chat-box")

        chatBox.innerHTML += `
          <div class="chat-message ${data.sender_type === "admin" ? "admin" : "user"}">
            <div class="chat-bubble">
              <div class="chat-sender">
                ${data.sender_type === "admin" ? "Event Manager" : (data.user_name || "You")}
              </div>
              <div class="chat-text">
                ${data.message}
              </div>
            </div>
          </div>
        `

        chatBox.scrollTop = chatBox.scrollHeight
      }
    }
  )
}

window.createQueryChannel = createQueryChannel

File: app/channels/query_channel.rb
class QueryChannel < ApplicationCable::Channel
  def subscribed
    @query = Query.find(params[:query_id])
    is_owner   = @query.user_id == current_user.id
    is_manager = @query.event.event_manager_id == current_user.id
    return reject_unauthorized_connection unless is_owner || is_manager || current_user.admin?
    stream_for @query
  end


  def send_message(data)
    message = @query.query_messages.create!(
      message: data['message'],
      sender_type: data['sender_type'],
      user: current_user
    )

    QueryChannel.broadcast_to(@query, {
      message: message.message,
      sender_type: message.sender_type,
      user_name: message.user.name
    })
  end
end
